### PR TITLE
feat: Add a new by_regex option.

### PR DESCRIPTION
**Pull request by JeffFaer**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -464,10 +464,77 @@ progress = (
...
+```diff
++# keep-sorted start by_regex=\b.+;\b
+ String bar;
+ Object baz;
+ List<String> foo;
+ # keep-sorted end
+```
+
+```diff
++# keep-sorted start by_regex=\b.+;\b prefix_order=foo
```
Comment (wachsler-google) [inline README.md]: It's not clear from the doc for prefix_order that `prefix_order=foo` is equivalent to `prefix_order=foo,`, i.e. that if no empty prefix is explicitly specified, one is assumed to be at the end. I can infer that from this example, but perhaps it's worth mentioning  explicitly

File: README.md [hunk 2]
```diff
@@ -464,10 +464,77 @@ progress = (
...
+
+> [!TIP]
+> Regular expressions often need special characters. See [Syntax](#syntax) below
+> for how to include special characters in the `by_regex` option.
+
+By default, all characters that the regular expression matches will be
+considered for sorting. If the regular expression contains any capturing groups,
+only the characters matched by the capturing groups will be considered for
+sorting. The result from each regular expression will be concatenated into a
+list of results, and that list of results will be sorted [lexicographically].
```
Comment (1e0ng) [inline README.md:482]: Just curious, do we allow users to use `by_regex` with `numeric`? Probably out of the scope of this change though.

File: goldens/by_regex.in [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  bool bar
+  int bar
+  keep-sorted-test end
+
+Cannot combine with ignore_prefixes
+  keep-sorted-test start by_regex=.* ignore_prefixes=foo,bar
+  2
+  1
+  3
+  keep-sorted-test end
```
Comment (1e0ng) [inline goldens/by_regex.in:40]: Shall we add some edge cases, for example, a regex can contain `,`, `'`, `[` or `]`.

**Review by 1e0ng — APPROVED** [review]
Thanks so much for this awesome change! Looking forward to this powerful feature in the new year.

File: README.md (hunk 2 above)
Comment (alzabo) [reply README.md:482]: The `numeric` and `by_regex` options interacted correctly for me when I tested a local build based on this patch. It seems like a good idea to note that the lexicographical sorting behavior can be modified by other options, and probably also to add a test case for the combination.

**Review by alzabo — APPROVED** [review]
LGTM. Thanks!

File: goldens/by_regex.in [hunk 4]
```diff
@@ -0,0 +1,40 @@
+No capturing group
```
Comment (KatrinaHoffert) [inline goldens/by_regex.in:1]: Another test case I'm particularly interested is a multi-line one (i.e., utilizing block or group mode). e.g., one FR we had before was sorting by function name when the function signatures might be something like:

```cpp
someReallyLongReturnType<CPPlol, IDK<whatever>>
  myFunc1();

int myFunc2();
```

Which can be done now, as regex would be able to ignore the line breaks and declare a capturing group to match the function name. (The ethics of trying to parse a C++ type with regex is omitted lol.)

File: keepsorted/options.go [hunk 5]
```diff
@@ -339,6 +359,35 @@ func (opts blockOptions) trimIgnorePrefix(s string) string {
...
+func (opts blockOptions) regexTransform(s string) []string {
+	if len(opts.ByRegex) == 0 {
+		return []string{s}
+	}
+
+	var ret []string
+	for _, regex := range opts.ByRegex {
+		m := regex.FindStringSubmatch(s)
+		if m == nil {
+			continue
```
Comment (KatrinaHoffert) [inline keepsorted/options.go:378]: Hmm, for the case of multiple regexes, is this really what we want, as opposed to, say, treating it as an empty string or some "always goes last" special value? Like, suppose I'm sorting a list of employees first by salary and then by how much I like their middle name. If I don't know their salary, I'd expect them to go at the top or bottom (kinda arbitrary) but then amongst all the people who didn't have salaries, I'd expect to still sort them correctly by how much I like their middle name.

By using `continue`, you instead end up with a nonsensical order because some lines are sorted by salary first while others are sorted by cool middle names despite the fact that salaries and bad-ass middle names are not comparable.

Since we lack a current special value for "sort last", the simplest thing to do would be to just change this to `ret = append(ret, ""); continue`, though putting non-matching last feels more intuitive.

File: keepsorted/options.go [hunk 6]
```diff
@@ -339,6 +359,35 @@ func (opts blockOptions) trimIgnorePrefix(s string) string {
...
+		}
+		if len(m) == 1 {
+			// No capturing groups. Consider all matched text.
+			ret = append(ret, m[0])
+		} else {
+			// At least one capturing group. Only consider the capturing groups.
+			ret = append(ret, m[1:]...)
+		}
+	}
+	return ret
```
Comment (KatrinaHoffert) [inline keepsorted/options.go:388]: Should we perhaps special case the event of `len(ret) == 0` (or none of them matched, if you do the above comment)? Because right now, if you have a couple of lines that fail to match any of your regexes, I'm unsure if the result is intuitive. You'd end up with all the non-matching lines sorted at the very end in a kinda arbitrary order (well, insertion order, cause stable sort).

We could perhaps just use the original, unchanged string instead. Honestly, non-matching is kinda weird, but I could see this as being a bit more intuitive than not sorting at all. Another slight variation would be to treat it like a `len(opts.ByRegex)`th element to the return value. That is, if normally we return values like `["int", "baz"]` and we encounter a non-matching line, we might return something like `["", "", "non-matching line"]`. That way it sorts above or below all the matching lines, but all the non-matching lines are sorted amongst themselves without any regex. I feel like that's pretty intuitive.

Alternatively, you could choose to make it a warning and emit a finding? Perhaps that would be best simply because it would avoid "bug" reports that are actually just user error from non-matching regexes. Though on the other hand, I'm not sure if we want to require regexes to have to match every line. I could see it being a feature to allow regexes to (sanely) not match every line, because that would let you declare some best effort sorting rules while still having a reasonable fallback without having to get your regex perfect. Having to tailor your regex to perfectly match every line could be toilsome in some cases.

File: goldens/by_regex.err [hunk 7]
```diff
@@ -0,0 +1 @@
+WRN by_regex cannot be used with ignore_prefixes (consider adding a non-capturing group to the start of your regex instead of ignore_prefixes: "(?:foo|bar)") line=36
```
Comment (KatrinaHoffert) [inline goldens/by_regex.err:1]: I forget offhand, is it even possible to cover the case of the regex failing to parse? That case panics, so perhaps it isn't feasible to test. But if it can be tested, we should do so. We can expect that users will hit that quite a lot, actually. It's a meme how much people mess up regexes lol.

**Review by KatrinaHoffert — APPROVED** [review]
Verrrry nice! I think this can actually resolve quite a few different issues that users have reported, thanks to how versatile regex is.

I kinda wonder if we should at some point support multi-line start stanzas, since as the options and their complexity grows, the `keep-sorted start` line is getting longer and longer.

File: README.md (hunk 1 above)
Comment (JeffFaer) [reply README.md]: I've updated the documentation for prefix sorting. Does that make it clearer?

File: goldens/by_regex.err (hunk 7 above)
Comment (JeffFaer) [reply goldens/by_regex.err:1]: It is possible! That case doesn't panic, it prints a warning on the offending line. Done

File: README.md (hunk 2 above)
Comment (JeffFaer) [reply README.md:482]: I've added more test cases. I added a test case for block=yes instead of group_prefixes, but in practice, group_prefixes should work just fine. It gets applied _before_ the regex, though. I've updated the docs to call that out.

File: goldens/by_regex.in (hunk 4 above)
Comment (JeffFaer) [reply goldens/by_regex.in:1]: Done!

File: keepsorted/options.go (hunk 6 above)
Comment (JeffFaer) [reply keepsorted/options.go:388]: > You'd end up with all the non-matching lines sorted at the very end in a kinda arbitrary order (well, insertion order, cause stable sort).

The `lessFn` uses `lineGroup.less` as a last resort, which compares the `joinedLines` content, and then the `comment` content, so they'll already get the ordering you're describing

----

I can see arguments both ways for whether we should emit a finding for regexes that don't match a line. I'm inclined to leave it as is for now, and if we start getting a lot of bug reports from users we can re-evaluate.

Comment (JeffFaer) [reply keepsorted/options.go:388]: Ah, I guess it's not _exactly_ the ordering you're describing since the other sorting options won't apply. But that seems like it might be fine since the user intended to sort the result of the regex in that configured way, and the non-matching lines might have different requirements

File: keepsorted/options.go (hunk 5 above)
Comment (JeffFaer) [reply keepsorted/options.go:379]: Good catch. I agree that this should have some sort of always goes last special value. That seems to be a bit more consistent with how `prefix_order` works (non matching lines go last)

PTAL

Comment (KatrinaHoffert) [reply keepsorted/options.go:379]: LGTM

**Review by KatrinaHoffert — APPROVED** [review]
Heh at those regex examples for function names. Should be interesting what monstrosities people will come up with trying to use regex to parse non-regular languages.

**Review by alzabo — APPROVED** [review]

**Review by 1e0ng — APPROVED** [review]
